### PR TITLE
ref: Add generic consumer CLI, introduce profiles consumer on it

### PR DESCRIPTION
**Pull request by untitaker**
We want to incrementally get rid of all consumer CLIs, and replace them with a
single consumer CLI that takes over all the operations boilerplate.

The most important part to me is that this new CLI already configures metrics
for you in a very opinionated way. All arroyo metrics go to `sentry.consumer`,
tagged by topic and consumer group, instead of being prefixed with random
stuff. This will allow for having consistent, reusable metric dashboards.

Those consumer-related tags are completely global, meaning that they are
available on custom metrics too. Additionally, they work out of the box in
multiprocessing if the RunTaskWithMultiprocessing class from
`sentry.utils.arroyo` is used. (we yet have to use and test that class, this PR
just prototypes it)

This migration has to be done one-by-one and on a new subcommand because some
of the existing consumers take different CLI arguments. We might use the
opportunity and just get rid of some CLI parameters, but there will have to be
some facility to register new CLI options per logical topic -- we can add this
feature later once we get to those more sophisticated consumers.

After this PR, we should update ops config to use the new CLI
`sentry run consumer profiles --consumer-group profiles`, then get rid of the `sentry run profiles` command.

### CODE REVIEW

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -2955,7 +2975,11 @@ def build_cdc_postgres_init_db_volume(settings):
     # Topic for indexer translated metrics
     KAFKA_SNUBA_METRICS: {"cluster": "default"},
     # Topic for receiving profiles from Relay
-    KAFKA_PROFILES: {"cluster": "default"},
+    KAFKA_PROFILES: {
+        "cluster": "default",
+        "group_id": "ingest-profiles",
+        "strategy_factory": "sentry.profiles.consumers.process.factory.ProcessProfileStrategyFactory",
```
Comment (lynnagara) [inline src/sentry/conf/server.py:2981]: I think it'd be better if this was hardcoded rather than in configuration. I don't think the user should be able to change this value, and it seems like it would be easy to accidentally clobber it due to the way we override the whole "KAFKA_PROFILES" object in getsentry/ops.

Comment (lynnagara) [reply src/sentry/conf/server.py:2981]: This structure also imposes a relationship between logical topics and consumers that we probably shouldn't. It should be possible to have 2 consumers connected to 1 logical topic, e.g. metrics indexer and last seen updater.

File: src/sentry/utils/arroyo.py [hunk 2]
```diff
@@ -51,3 +62,105 @@ def timing(self, name: str, value: Union[int, float], tags: Optional[Tags] = Non
...
+    _initialize_arroyo_main()
+
+
+def _initialize_arroyo_main() -> None:
+    from arroyo import configure_metrics
+
+    from sentry.utils.metrics import backend
+
+    metrics_wrapper = MetricsWrapper(backend, name="consumer")
+    configure_metrics(metrics_wrapper)
```
Comment (lynnagara) [inline src/sentry/utils/arroyo.py:101]: Should we skip this in the subprocess initialization? Since Arroyo never fires metrics from there it doesn't do anything.

File: src/sentry/utils/arroyo.py [hunk 3]
```diff
@@ -51,3 +62,105 @@ def timing(self, name: str, value: Union[int, float], tags: Optional[Tags] = Non
...
+
+
+class RunTaskWithMultiprocessing(ProcessingStrategy[Union[FilteredPayload, TStrategyPayload]]):
```
Comment (lynnagara) [inline src/sentry/utils/arroyo.py:104]: I'd probably do this in a separate PR since it can already be used in existing consumers without migrating over to the unified CLI. But up to you.

Comment (untitaker) [reply src/sentry/utils/arroyo.py:102]: yeah it's completely separate, but i had to wrap my head around all the initialization e2e and figured the PR isn't quite that big anyway

File: src/sentry/conf/server.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/conf/server.py]: yeah I realized the same. I changed it so that consumers are now registered by their consumer group instead.

File: src/sentry/runner/commands/run.py [hunk 4]
```diff
@@ -85,7 +86,7 @@ def inner(f):
             "--consumer-group",
             "group_id",
             default=consumer_group,
-            help="Kafka consumer group for the consumer.",
+            help="(physical) Kafka consumer group for the consumer.",
```
Comment (lynnagara) [inline src/sentry/runner/commands/run.py:89]: Does this mean there is now a concept being introduced of a separate logical and physical consumer group name?

File: src/sentry/conf/server.py (hunk 1 above)
Comment (lynnagara) [reply src/sentry/conf/server.py]: I'm not sure we should be registering by consumer group either. Consumer group is something that gets passed via the command line and can change frequently (e.g. during incidents). I think they should be hardcoded types with some other key that identifies them. In Snuba we use the concept of "storage" here where each writeable storage is associated with one consumer.

Comment (untitaker) [reply src/sentry/conf/server.py]: Yeah i was thinking we could have a concept like a "logical consumer group". The physical can still be overridden in this pr

File: src/sentry/consumers/__init__.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    strategy_factory: str
+
+
+# consumer group -> consumer definition
+KAFKA_CONSUMERS: Mapping[str, ConsumerDefinition] = {
+    "ingest-profiles": {
+        "topic": settings.KAFKA_PROFILES,
+        "strategy_factory": "sentry.profiles.consumers.process.factory.ProcessProfileStrategyFactory",
+    }
+}
```
Comment (lynnagara) [inline src/sentry/consumers/__init__.py:17]: This structure seems good to me though let's not refer to the key here as a "consumer group" which means something very different in Kafka? How about we call this like a "Sentry consumer name" or something else less overloaded?

File: src/sentry/utils/arroyo.py [hunk 6]
```diff
@@ -51,3 +62,103 @@ def timing(self, name: str, value: Union[int, float], tags: Optional[Tags] = Non
...
+
+    topic = force_topic or topic
+
+    cluster_name: str = force_cluster or settings.KAFKA_TOPICS[topic]["cluster"]
+    consumer_config = build_kafka_consumer_configuration(
+        kafka_config.get_kafka_consumer_cluster_options(
+            cluster_name,
+        ),
+        group_id=group_id,
+        auto_offset_reset="latest",
```
Comment (lynnagara) [inline src/sentry/utils/arroyo.py:150]: is this to be swapped out? i guess we should also apply the strict offset reset setting passed in?

File: src/sentry/utils/arroyo.py [hunk 7]
```diff
@@ -51,3 +62,103 @@ def timing(self, name: str, value: Union[int, float], tags: Optional[Tags] = Non
...
+    force_cluster: str | None,
+    strategy_factory_cls: Type[ProcessingStrategyFactory[Any]],
+) -> None:
+    from django.conf import settings
+
+    from sentry.metrics.middleware import add_global_tags
+
+    add_global_tags(kafka_topic=topic, group_id=group_id)
+
+    topic = force_topic or topic
```
Comment (lynnagara) [inline src/sentry/utils/arroyo.py:142]: I know this existed before but having `topic` and `force_topic` in this API is super confusing

File: src/sentry/runner/commands/run.py (hunk 4 above)
Comment (untitaker) [reply src/sentry/runner/commands/run.py:89]: Yeah that's sort of what i was going for, in the absence of just creating another term. But I'm fine with renaming this concept to something else.

File: src/sentry/utils/arroyo.py (hunk 7 above)
Comment (untitaker) [reply src/sentry/utils/arroyo.py:142]: I just realized that force_topic still needs to be a valid logical topic, so now I wonder what the use of that is. I simplified the logic of force_topic a bit based on that.

File: src/sentry/consumers/__init__.py (hunk 5 above)
Comment (untitaker) [reply src/sentry/consumers/__init__.py:17]: I renamed consumer group to consumer name

**Review by lynnagara — APPROVED** [review]
